### PR TITLE
Updates to allow retrieving a token using the device flow on Salesforce.

### PR DESCRIPTION
**Pull request by markcatley**
For discussion. Probably needs a bit more thought before finalising.

### CODE REVIEW

**Comment by markcatley** [comment]
`("grant_type", "http://oauth.net/grant_type/device/1.0"),` doesn't work (needs to be `grant_type`) either but it can be overridden in the secret file using a query parameter.

File: src/device.rs [hunk 1]
```diff
@@ -218,12 +221,17 @@ where
                             return Err(RequestError::ClientError(err));
                         }
                         Ok(res) => {
+                            // This return type is defined in https://tools.ietf.org/html/draft-ietf-oauth-device-flow-07#section-3.2
+                            // The alias is present as Google use a non-standard name for verification_uri.
```
Comment (dermesser) [inline src/device.rs:225]: nasty!

**Review by dermesser — APPROVED** [review]
I checked it out locally, ran the examples as well as tests, and it looks good! The code itself too, of course. It's a good idea to do more logging as well, I've completely neglected that so far.

**Comment by dermesser** [comment]
> `("grant_type", "http://oauth.net/grant_type/device/1.0"),` doesn't work (needs to be `grant_type`) either but it can be overridden in the secret file using a query parameter.

Can you elaborate on this? I don't quite understand the problem you're stating; is the query parameter name wrong?

File: src/device.rs (hunk 1 above)
Comment (markcatley) [reply src/device.rs:225]: Yup! Welcome to software!

I've updated this link as that's an old version of the spec. Of course, it's still draft too!

**Comment by markcatley** [comment]
> I checked it out locally, ran the examples as well as tests, and it looks good! The code itself too, of course. It's a good idea to do more logging as well, I've completely neglected that so far.

**Comment by markcatley** [comment]
> > `("grant_type", "http://oauth.net/grant_type/device/1.0"),` doesn't work (needs to be `grant_type`) either but it can be overridden in the secret file using a query parameter.
> 
> Can you elaborate on this? I don't quite understand the problem you're stating; is the query parameter name wrong?

This is more of an interesting (or nasty) note as I'm able to work around it using the token_uri int he secret file. I should go and have a look at what other OAuth2 implementations do here. If every provider requires something slightly different they must have some way of handling it. I should possibly add something to the documentation so that others can get this working.

Anyhow, this is the section of the spec:
[Device Flow Spec Section 3.4](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.4)
```
grant_type
      REQUIRED.  Value MUST be set to
      "urn:ietf:params:oauth:grant-type:device_code".
```
Neither Google nor Salesforce suggests that token. Salesforce definitely doesn't work with that (or what Google requires - it needs "device_code", I haven't checked Google yet.

**Comment by dermesser** [comment]
Google seems to want to see that URL: https://developers.google.com/identity/protocols/OAuth2ForDevices#step-4-poll-googles-authorization-server

```
grant_type | Set this value to http://oauth.net/grant_type/device/1.0.
```

**Comment by dermesser** [comment]
I've uploaded this to crates.io as v3.0.2-beta.

**Pull request merged**